misk/src/main/kotlin/misk/web/extractors/FormValueParameterExtractorFactory.kt

```diff
@@ -0,0 +1,118 @@
+package misk.web.extractors
+
+import com.google.common.collect.LinkedHashMultimap
+import com.google.common.collect.Multimap
+import misk.web.FormValue
+import misk.web.PathPattern
+import misk.web.Request
+import misk.web.actions.WebAction
+import okio.BufferedSource
+import java.net.URLDecoder
+import java.util.regex.Matcher
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+
+object FormValueParameterExtractorFactory : ParameterExtractor.Factory {
+    /**
+     * Creates a [ParameterExtractor] that extracts the form value parameter with the same name as
+     * [parameter] and returns it as whatever type is specified in [parameter]. Returns null if the
+     * parameter doesn't occur in the request, or can't be parsed into the correct type.
+     */
+    override fun create(
+            function: KFunction<*>,
+            parameter: KParameter,
+            pathPattern: PathPattern
+    ): ParameterExtractor? {
+        parameter.findAnnotation<FormValue>() ?: return null
+        if (parameter.type.classifier !is KClass<*>) return null
+
+        val kClass = parameter.type.classifier as KClass<*>
+        val constructor = kClass.primaryConstructor ?: return null
+        val constructorParameters = constructor.parameters.map {
+            val isList = it.type.classifier?.equals(List::class) ?: false
+            ConstructorParameter(
+                    it,
+                    it.name,
+                    it.isOptional,
+                    it.type.isMarkedNullable,
+                    isList,
+                    converterFor(if (isList) it.type.arguments.first().type!! else it.type)
+            )
+        }
+
+        return object : ParameterExtractor {
+            override fun extract(
+                    webAction: WebAction,
+                    request: Request,
+                    pathMatcher: Matcher
+            ): Any? {
+                val formValueMapping = parseBody(request.body)
+                val parameterMap = LinkedHashMap<KParameter, Any?>()
+                for (p in constructorParameters) {
+                    val parameterValue = formValueMapping[p.name]
```

Comment:
form field names are case insensitive, so need to force both the parsed field name and the name that's in the `ConstructorParameter` to be lower case

Reply:
yep done